File: backend/src/main/java/com/ecommerce/app/service/ProductService.java
package com.ecommerce.app.service;

import com.ecommerce.app.model.Product;
import com.ecommerce.app.repository.ProductRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.UUID;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {
    private final ProductRepository productRepository;

    public List<Product> getAllProducts(String query) {
        if (query != null && !query.isEmpty()) {
            return productRepository.searchProducts(query);
        }
        return productRepository.findAll();
    }

    public Product getProductById(UUID id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Product not found with id: " + id));
    }

    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategory(category);
    }

    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    public Product updateProduct(UUID id, Product productDetails) {
        Product product = getProductById(id);
        product.setTitle(productDetails.getTitle());
        product.setDescription(productDetails.getDescription());
        product.setPrice(productDetails.getPrice());
        product.setCategory(productDetails.getCategory());
        product.setImage(productDetails.getImage());
        product.setRating(productDetails.getRating());
        return productRepository.save(product);
    }

    public void deleteProduct(UUID id) {
        Product product = getProductById(id);
        productRepository.delete(product);
    }
}

File: frontend/src/services/ProductService.js
import axios from 'axios';

const BASE_URL = 'http://localhost:8080/api/products';

const ProductService = {
    getAllProducts: async (query = '') => {
        try {
            const response = await axios.get(`${BASE_URL}${query ? `?query=${query}` : ''}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching products:', error);
            throw error;
        }
    },

    getProductById: async (id) => {
        try {
            const response = await axios.get(`${BASE_URL}/${id}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching product with id ${id}:`, error);
            throw error;
        }
    },

    getProductsByCategory: async (category) => {
        try {
            const response = await axios.get(`${BASE_URL}/category/${category}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching products in category ${category}:`, error);
            throw error;
        }
    },

    createProduct: async (productData) => {
        try {
            const response = await axios.post(BASE_URL, productData);
            return response.data;
        } catch (error) {
            console.error('Error creating product:', error);
            throw error;
        }
    },

    updateProduct: async (id, productData) => {
        try {
            const response = await axios.put(`${BASE_URL}/${id}`, productData);
            return response.data;
        } catch (error) {
            console.error(`Error updating product with id ${id}:`, error);
            throw error;
        }
    },

    deleteProduct: async (id) => {
        try {
            await axios.delete(`${BASE_URL}/${id}`);
        } catch (error) {
            console.error(`Error deleting product with id ${id}:`, error);
            throw error;
        }
    }
};

export default ProductService;
